lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp

```diff
@@ -433,18 +438,31 @@ Value loadA(ConversionPatternRewriter &rewriter, Location loc, Value thread,
 
     Type smemPtrTy = getShemPtrTy(aElemTy);
 
+    int loadsPerThread = offsets.size() / (numRepM * numRepK);
+    const int elemsPerLoad = numOfElems / loadsPerThread;
+    assert(numOfElems % loadsPerThread == 0);
+
     for (int m = 0; m < numRepM; ++m) {
       for (int k = 0; k < numRepK; ++k) {
         auto vecTy = vec_ty(aElemTy, numOfElems);
         Value valVec = undef(vecTy);
-        for (unsigned elem = 0; elem < numOfElems; ++elem) {
-          Value elemOffset =
-              offsets[m * numOfElems * numRepK + k * numOfElems + elem];
-          Value elemValue = load(gep(smemPtrTy, smemBase, elemOffset));
-          if (numOfElems > 1)
-            valVec = insert_element(vecTy, valVec, elemValue, i32_val(elem));
-          else
-            valVec = elemValue;
+        for (unsigned loadId = 0; loadId < loadsPerThread; ++loadId) {
+          auto loadVecTy = vec_ty(aElemTy, elemsPerLoad);
+          Value loadOffset =
+              offsets[m * loadsPerThread * numRepK + k * loadsPerThread + loadId];
+          Value loadAddress = bitcast(gep(smemPtrTy, smemBase, loadOffset),
+                                      getShemPtrTy(loadVecTy));
+          Value vectorValue = load(loadAddress);
+          if (numOfElems > 1) {
+            for (int elemId = 0; elemId < elemsPerLoad; ++elemId) {
```

Comment:
Here we implicitly assume `elemsPerLoad` = `numOfElems`, i.e. the vector load length is the same as the number of elements per thread for a **single** mfma instruction. 
- If we want to generalize the vector load in the fast path, we need to consider the case when `elemsPerLoad` < `numOfElems`.
- Or if we just want to make the fast path "fast", we'll always have `loadsPerThread` = 1 and we can further simply the code in the fast path.

I'd personally go with the second one since ultimately we'll enable swizzling and make the normal path faster and faster.

Reply:
I don't understand why we want smaller vector loads here.
With 32x32x32 mfma instruction largest `numOfElems` is 4.

Maybe we understand "fast path" differently?
By this term I mean an algorithm which generates better code for some specific case,
"normal" path is a general algorithm that generates which works for all cases, but could be slower or consume more resources. 

Let's discuss this and make another PR if needed.